### PR TITLE
[SPARK-33382][SQL][TESTS] Unify datasource v1 and v2 SHOW TABLES tests

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to gather common `SHOW TABLES` tests into one trait `org.apache.spark.sql.execution.command.ShowTablesSuite`, and put datasource specific tests to the `v1.ShowTablesSuite` and `v2.ShowTablesSuite`. Also tests for parsing `SHOW TABLES` are extracted to `ShowTablesParserSuite`.

### Why are the changes needed?
- The unification will allow to run common `SHOW TABLES` tests for both DSv1 and DSv2
- We can detect missing features and differences between DSv1 and DSv2 implementations.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
By running new test suites:
- `org.apache.spark.sql.execution.command.v1.ShowTablesSuite`
- `org.apache.spark.sql.execution.command.v2.ShowTablesSuite`
- `ShowTablesParserSuite`

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon Please, take a look at this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    withTable("source", "source2") {
+      val df = spark.createDataFrame(Seq((1L, "a"), (2L, "b"), (3L, "c"))).toDF("id", "data")
+      df.createOrReplaceTempView("source")
+      val df2 = spark.createDataFrame(Seq((4L, "d"), (5L, "e"), (6L, "f"))).toDF("id", "data")
+      df2.createOrReplaceTempView("source2")
+      f
+    }
+  }
+
+  test("show tables") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:51]: Can we put the parser tests in `ShowTablesParserSuite`? It doesn't involve data source and doesn't need to be run twice with DS v1 and v2.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+package org.apache.spark.sql.execution.command.v1
+
+import org.apache.spark.sql.{AnalysisException, Row}
+import org.apache.spark.sql.catalyst.analysis.NoSuchDatabaseException
+import org.apache.spark.sql.execution.command.{ShowTablesSuite => CommonShowTablesSuite}
+import org.apache.spark.sql.types.{BooleanType, StringType, StructType}
+
+class ShowTablesSuite extends CommonShowTablesSuite {
+  override def catalog: String = "spark_catalog"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala:26]: let's not hardcode it. `CatalogManager.SESSION_CATALOG_NAME`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      .add("tableName", StringType, nullable = false)
+      .add("isTemporary", BooleanType, nullable = false)
+  }
+  override protected def getRows(showRows: Seq[ShowRow]): Seq[Row] = {
+    showRows.map {
+      case ShowRow(namespace, table, isTemporary) => Row(namespace, table, isTemporary)
+    }
+  }
+
+  // `SHOW TABLES` returns empty result in V2 catalog instead of throwing the exception.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala:40]: this is something we should fix later.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:51]: Do you mean in a separate file or a class in the same file?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+import org.apache.spark.sql.{QueryTest, Row}
+import org.apache.spark.sql.catalyst.analysis.{AnalysisTest, UnresolvedNamespace}
+import org.apache.spark.sql.catalyst.parser.CatalystSqlParser.parsePlan
+import org.apache.spark.sql.catalyst.plans.logical.{ShowTables, ShowTableStatement}
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{BooleanType, StringType, StructType}
+
+trait ShowTablesSuite extends QueryTest with SharedSparkSession with AnalysisTest {
+  protected def catalog: String
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:28]: A simpler way to test v1 vs v2 is: In v2 test, register a v2 catalog and set it as the default. Then in the test cases we just need to write `$namespace.$table`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:51]: In a separate file is better.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala (hunk 4 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:28]: yeh, I thought about this but wanted to avoid dependency from a global variable. Also it will make harder to write tests that:
- check `USE ...`
- check both dsv1 and dsv2 in the same test

And with explicit catalog, it should be easier to mix the common trait to other test suite like JDBC or Hive catalogs.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala:40]: I will open an JIRA for this, and add `TODO` here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:28]: OK makes sense.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala:40]: Here is the umbrella ticket for TODOs: https://issues.apache.org/jira/browse/SPARK-33392

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesParserSuite.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      ShowTables(UnresolvedNamespace(Seq("ns1")), Some("*test*")))
+    comparePlans(
+      parsePlan("SHOW TABLES IN ns1 LIKE '*test*'"),
+      ShowTables(UnresolvedNamespace(Seq("ns1")), Some("*test*")))
+  }
+
+  test("show table extended") {
+    comparePlans(
+      parsePlan("SHOW TABLE EXTENDED LIKE '*test*'"),
+      ShowTableStatement(None, "*test*", None))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesParserSuite.scala:55]: @imback82 do you know why we still have `ShowTableStatement` here?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+        runShowTablesSql(
+          s"SHOW TABLES FROM $catalog.db LIKE '*2'",
+          Seq(ShowRow("db", "table_name_2", false)))
+      }
+    }
+  }
+
+  // TODO(SPARK-33393): Support SHOW TABLE EXTENDED in DSv2
+  test("SHOW TABLE EXTENDED for default") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:95]: this test assumes that the default catalog is v1 catalog. Shall we move it to `v1.ShowTablesSuite`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  test("show table in a not existing namespace") {
+    val msg = intercept[NoSuchDatabaseException] {
+      runShowTablesSql(s"SHOW TABLES IN $catalog.unknown", Seq())
+    }.getMessage
+    assert(msg.contains("Database 'unknown' not found"))
+  }
+
+  // `SHOW TABLES` from v2 catalog returns empty result.
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala:49]: How is this related to v2? This test checks the default catalog, which by default is v1.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  test("using v2 catalog with empty namespace") {
+    withTable(s"$catalog.table") {
+      spark.sql(s"CREATE TABLE $catalog.table (id bigint, data string) $defaultUsing")
+      runShowTablesSql(s"SHOW TABLES FROM $catalog", Seq(ShowRow("", "table", false)))
+    }
+  }
+
+  // The test fails for V1 catalog with the error:
+  // org.apache.spark.sql.AnalysisException:
+  //   The namespace in session catalog must have exactly one name part: spark_catalog.table
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:75]: If we want to focus on the default catalog, let's use one-level namespace to have a fair comparison

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      testV1CommandNamespace("SHOW TABLE EXTENDED " +
+        s"FROM $namespace LIKE 'tb*' PARTITION(id=1)",
+        namespace)
+      testV1CommandNamespace("SHOW TABLE EXTENDED " +
+        s"IN $namespace LIKE 'tb*' PARTITION(id=1)",
+        namespace)
+    }
+  }
+
+  // The test fails with the error in V1 session catalog:
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:116]: Again, to have a fail comparison, let's use one leve namespace here.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala (hunk 8 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:75]: How about to change the condition here: https://github.com/apache/spark/blob/45b7212fd3d60f537aa034a6255188e0283d05e7/sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/LookupCatalog.scala#L64 and don't throw the exception if `ident.namespace.length == 2 && ident.namespace(0) == "spark_catalog"`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesParserSuite.scala (hunk 5 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesParserSuite.scala:55]: We parse `SHOW TABLE EXTENDED`:
https://github.com/apache/spark/blob/ccc0250a08fa9519a6dc3c1153e51c1e110f1d7d/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4#L211-L212
and convert it to `ShowTableStatement `:
https://github.com/apache/spark/blob/0ad35ba5f8bd6413669b568de659334bb9a3fb44/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala#L2929-L2934

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala (hunk 6 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:95]: DSv2 doesn't support `SHOW TABLE EXTENDED`. After https://issues.apache.org/jira/browse/SPARK-33393, the test can work for DSv2 too with minimal changes:
- Abstract the column name `database` - in DSv2 we should show `namespace` probably
- DSv2 will not show views (at least at the moment).

I would split the test to 2 tests:
- SHOW TABLE EXTENDED for tables
- SHOW TABLE EXTENDED for views.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala (hunk 7 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala:49]: I described the reason why the test is placed to V1 specific test suite - because V2 doesn't show views otherwise we could move it to the common trait and set `default` to a V2 catalog.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:95]: My point is that: the tests in the base trait, should work for both v1 and v2 catalogs, and will be run with both v1 and v2 catalog.

The test now just run twice repeatedly in `v1.ShowTablesSuite` and `v2.ShowTablesSuite`, which doesn't make sense.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala (hunk 7 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/ShowTablesSuite.scala:49]: The confusion is: temp views don't belong to catalog, but belong to the spark session. Can we create permanent views in this test? Or change the test name to `v1  SHOW TABLES` list the temp views.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala (hunk 8 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:75]: We can't do that. v1 catalog can't support 2 level namespace, even if the first part name is "spark_catalog". It's still the namespace, not a catalog name, in that context.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesParserSuite.scala (hunk 5 above)
Comment (imback82) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesParserSuite.scala:55]: @cloud-fan `ShowTable` is not migrated yet; only `ShowTables` with "s" is migrated. I will work on this migration soon.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala (hunk 6 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/ShowTablesSuite.scala:95]: Create 2 separate tests for dsv1 and dsv2

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala (hunk 9 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:116]: What do you propose here? The original test has nested namespaces, BTW.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala (hunk 8 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:75]: I split the test, and opened the JIRA: https://issues.apache.org/jira/browse/SPARK-33403

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+      spark.sql(s"CREATE TABLE $catalog.table (id bigint, data string) $defaultUsing")
+      runShowTablesSql(s"SHOW TABLES FROM $catalog", Seq(ShowRow("", "table", false)))
+    }
+  }
+
+  test("namespace is not specified and the default catalog is set") {
+    withSQLConf(SQLConf.DEFAULT_CATALOG.key -> catalog) {
+      withTable("table") {
+        spark.sql(s"CREATE TABLE table (id bigint, data string) $defaultUsing")
+        // TODO(SPARK-33403): DSv2 SHOW TABLES doesn't show `default`
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:78]: This doesn't need TODO as it's a corrected behavior.

`CREATE TABLE table` creates a table in the current catalog and current namespace. The current namespace, if hasn't been set by users via USE yet, is decided by the current catalog. For the v1 catalog, the default current  namespace is `["default"]`. For the testing v2 catalog, it's `[]`.

I think a better way to write this test is:
```
spark.sql(s"CREATE TABLE $catalog.table (id bigint, data string) $defaultUsing")
runShowTablesSql("SHOW TABLES", Seq(ShowRow("", "table", false)))
```

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/v2/ShowTablesSuite.scala:78]: The code above returns the same in DSv2 but fails in DSv1:
```
  test("namespace is not specified and the default catalog is set") {
    withSQLConf(SQLConf.DEFAULT_CATALOG.key -> catalog) {
      withTable("table") {
        spark.sql(s"CREATE TABLE $catalog.table (id bigint, data string) $defaultUsing")
        runShowTablesSql("SHOW TABLES", Seq(ShowRow("", "table", false)))
      }
    }
  }
```
```
The namespace in session catalog must have exactly one name part: spark_catalog.table;
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
